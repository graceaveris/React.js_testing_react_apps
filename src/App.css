.App {
  text-align: center;
  font-family: "Century Gothic", CenturyGothic;
  background-color: #0D0C12;
  min-height: 100vh;
  display: flex;
  justify-content: center;
  text-align: center;
  flex-direction: column;
  font-size: calc(10px + 2vmin);
  padding-bottom: 100px;
  color: white;

}

.App-logo {
  animation: App-logo-spin infinite 20s linear;
  height: 20vmin;
  float: left
}

.App-header {
  background-color: #0D0C12;
  margin-top: 80px;
  min-height: 10vh;
  display: flex;
  flex-direction: column;
  align-items: center;
  font-size: calc(10px + 2vmin);
  color: white;
}

.main-btn {
  padding: 10px;
}

button {
  width: 150px;
  height: 40px;
  border-radius: 0px;
  border: none;
  margin: 5px;
  font-size: 20px;
  color: white;
  background-color: #8a83c6;
}

input {
  width: 250px;
  height: 25px;
  border-radius: 0px;
  margin: 5px;
  font-size: 20px;
  padding: 5px;
}

h1 {
  color: #8a83c6;
  font-size: calc(10px + 6vmin);
}

h2 {
  color: white;
  font-size: calc(10px + 2vmin);
}

p {
  font-size: calc(6px + 2vmin); 
}

li {
  font-size: 18px;
  list-style: none;
  padding: 3px;
}