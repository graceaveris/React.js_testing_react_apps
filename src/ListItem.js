import React from 'react';
const ListItem = ( props ) => {

return (

  <div>
    <p>{props.item}</p>
    <button className="smallbtn" onClick={() => props.onClick(props.id)}>delete</button>  
  </div>

  );
};

export default ListItem;
