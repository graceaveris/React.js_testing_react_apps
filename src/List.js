import React, { Component } from 'react';
import ListItem from './ListItem';

class List extends Component {


render () {

let items = (
       <div>
        {this.props.items.map((item) => {
          return <ListItem
          item={item.name}
          id={item.key}
          key={item.key}
          onClick={this.props.deleteItem}/>
        })} 
      </div> 
    );

 return (

    <div>
      <ul className="list">
       {items}
      </ul>
    </div>
 
  )};
};


export default List;
