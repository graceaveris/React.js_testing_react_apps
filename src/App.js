  import React, { Component } from 'react';
  import './App.css';
  import List from './List';

  class App extends Component {

    state = {
      count: 2,
      newItem: '',
      items: [
      { name: 'apple', key: 1 },
      { name: 'milk', key: 2 }
      ] 
    }

    addItem = () => {
      if (this.state.newItem) {

        let itemToAdd = {}
        itemToAdd.name = this.state.newItem
        itemToAdd.key = Math.random()

     this.setState({ count: this.state.count + 1, items: [...this.state.items, itemToAdd], newItem: '' })
     } 
    }

    deleteItem = (key) => {
      let items = [...this.state.items]
      let newItemsArray = []
      items.forEach((item) => {
        if (item.key !== key) { 
          newItemsArray.push(item)}
        })
     this.setState ({ items: newItemsArray, count: this.state.count - 1 })
    }

    render() {

    return(
    <div className="App">
       
       <header className="App-header">
        <h1>30 days of React</h1>
        <h2>Day Thirty / Testing React Apps</h2>
        <p>Total Items: {this.state.count}</p>
       </header>
       
       <div className="container">
        <input type='text' value={this.state.newItem} onChange={e => this.setState({ newItem: e.target.value})}></input>
        <button className='main-btn' onClick={this.addItem}>Add To List</button>
       </div>
      
      <div>
        <List items={this.state.items} deleteItem={this.deleteItem} />
      </div>

      </div>

      ) 
     }
   }

  export default App;
