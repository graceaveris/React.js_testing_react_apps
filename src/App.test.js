import React from 'react';
import ReactDOM from 'react-dom';

import { shallow } from 'enzyme';
import { render } from '@testing-library/react';
import renderer from 'react-test-renderer';

import App from './App';
import List from './List';
import ListItem from './ListItem';



//SHALLOW TESTS or 'unit tests' - Jest + enzyme

// this is a checking that our app loads
describe ('App component - load test', () => {
	it('renders without crashing', () => {
		shallow(<App />);
	});
});

//we describe which component we are going to test
describe('App component - units tests - jest/enzyme', () => {
    
    // unit test, checking count of 0 displaying on first render
	it('starts count at 2', () => {
		const wrapper = shallow(<App />);
		const text = wrapper.find('p').text();
		expect(text).toEqual('Total Items: 2')
	});
    
	// unit test, testing user interaction by simulating a click
	it('increments count by 1 on click', () => {
		const wrapper = shallow(<App />);

		const input = wrapper.find('input')
		input.simulate('change', {target: {value: 'test'} })
		
		const button = wrapper.find('button.main-btn');
		button.simulate('click');
		
		const text = wrapper.find('p').text();
		expect(text).toEqual('Total Items: 3');
	});
    
    // snapshot, snapshot of your component tree is created first run
    // then if it changes during dev you will be alerted
    // note: select u during test fail to update the snapshot
	it('matches snapshot', () => {
		const tree = renderer.create(<App />).toJSON();
		expect(tree).toMatchSnapshot();
	}); 

});


// MOCK TEST
// we create a fake function to test flow, or check for function calls 

	//define empty mock function to simulate the click 
	const mockClick = jest.fn();

	describe('ListItem component - mock test - jest/enzyme', () => {
		it('button click should fire function', () => {
			const component = shallow(<ListItem onClick={mockClick} />);

			const button = component.find('button.smallbtn');
			button.simulate('click');
			expect(mockClick).toHaveBeenCalled();
		});
	}); 


// TEST w REACT-TESING-LIBRARY

// INTEGRATION TEST
// using react-testing-lib to ckeck two components working together
describe('List component - integration test - react testing lib', () => {
	test('It renders shopping list items with delete buttons', () => {
		
		//we add our dummy data
		const item1 = {
			name: 'potato',
			key: 1
		}

		const item2 = {
			name: 'cheese',
			key: 2
		}
		//arrange data as props in correct format for component
		const props = {
			items:[ item1, item2 ]
		}
	
    //render data
   const { getByText, getAllByText } = render(<List {...props} />)

    //assert results
    const firstListItemNode = getByText(item1.name)
    const secondListItemNode = getByText(item2.name)
    const buttonCount = getAllByText('delete').length

    expect(firstListItemNode).toBeDefined()
    expect(firstListItemNode).toBeDefined()
    expect(buttonCount).toBe(2)

	})
})


// TERMINOLOGY
// expect is an assertion
// .toEqual is a matcher
// shallow means its only testing component provided, ignoring children!
// unit test - tests one function or component - shallow()
// integration test - testing a react component tree - mount()

