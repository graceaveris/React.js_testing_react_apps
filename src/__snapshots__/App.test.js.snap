// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App component - units tests - jest/enzyme matches snapshot 1`] = `
<div
  className="App"
>
  <header
    className="App-header"
  >
    <h1>
      30 days of React
    </h1>
    <h2>
      Day Thirty / Testing React Apps
    </h2>
    <p>
      Total Items: 
      2
    </p>
  </header>
  <div
    className="container"
  >
    <input
      onChange={[Function]}
      type="text"
      value=""
    />
    <button
      className="main-btn"
      onClick={[Function]}
    >
      Add To List
    </button>
  </div>
  <div>
    <div>
      <ul
        className="list"
      >
        <div>
          <div>
            <p>
              apple
            </p>
            <button
              className="smallbtn"
              onClick={[Function]}
            >
              delete
            </button>
          </div>
          <div>
            <p>
              milk
            </p>
            <button
              className="smallbtn"
              onClick={[Function]}
            >
              delete
            </button>
          </div>
        </div>
      </ul>
    </div>
  </div>
</div>
`;
